qcodes/utils/dataset/doNd.py

```diff
@@ -483,6 +483,180 @@ def do2d(
+
+    def _make_nested_setpoints(
+        params: List[_BaseParameter],
+        param_dicts: Dict[_BaseParameter, Dict[str, Union[float, int]]]
+    ) -> Union[np.ndarray, List[Tuple[()]]]:
+        """Create the cartesian product of all the setpoint values."""
+
+        setpoint_values = [np.linspace(params_dicts[p]['start'],
+                                       params_dicts[p]['stop'],
+                                       int(params_dicts[p]['num_points']))
+                           for p in params]
+        setpoint_grids = np.meshgrid(*setpoint_values, indexing='ij')
+        flat_setpoint_grids = [np.ravel(grid, order='C') for grid in setpoint_grids]
+        flat_setpoints: Union[np.ndarray, List[Tuple[()]]]
+        if flat_setpoint_grids:
+            flat_setpoints = np.vstack(flat_setpoint_grids).T
+        else:
+            flat_setpoints = [()]
+        return flat_setpoints
+
+    def _find_parameters(*params: Union[_BaseParameter, int, float, ParamMeasT]
+                         ) -> Tuple[List[Union[_BaseParameter, ParamMeasT]],
+                                    Dict[Union[_BaseParameter, ParamMeasT], List[Union[int, float]]]]:
+        _params: List[Union[_BaseParameter, ParamMeasT]] = []
+        args_per_parameter: Dict[Union[_BaseParameter, ParamMeasT], List[Union[int, float]]] = {}
+        for par in params:
+            if isinstance(par, _BaseParameter) or callable(par):  # A QCodes parameter:
+                _params.append(par)
+                args_per_parameter[par] = []
+                last_param = par
+            else:  # A numerical argument:
+                try:
+                    float(par)
+                    args_per_parameter[last_param].append(par)
+                except TypeError as err:
+                    raise TypeError(f'Invalid argument type: {par}') from err
+        return _params, args_per_parameter
+
+    def _parse_dond_arguments(*params: Union[_BaseParameter,
+                              int, float, ParamMeasT]
+                              ) -> Tuple[
+        List[_BaseParameter],
+        Dict[_BaseParameter, Dict[str, Union[float, int]]],
+        List[ParamMeasT]
+    ]:
+        """
+        Parse arguments for a dond scan.
+
+        The expected order for the arguments is:
+        params: param_set_1, start_1, stop_1, num_points_1, delay_1, ...,
+                param_set_n, start_n, stop_n, num_points_1, delay_n,
+                param_meas_1, param_meas_2, ..., param_meas_m
+        """
+        _params, args_per_parameter = _find_parameters(*params)
+        arg_types = ['start', 'stop', 'num_points', 'delay']
+        params_set: List[_BaseParameter] = []
+        params_meas: List[ParamMeasT] = []
+        params_dicts: Dict[_BaseParameter, Dict[str, Union[float, int]]] = {}
+        for param in _params:
+            arguments = args_per_parameter[param]
+            if not arguments:
+                if isinstance(param, _BaseParameter):
```

Comment:
this if statement might be replaced with an assert statement like ` assert isinstance(param, _BaseParameter)` as we are making sure that param is an instance of _BaseParameter as that is expected